### PR TITLE
Improved LOR solvers, including for H(curl) and H(div) [lor-hcurl-hdiv]

### PR DESCRIPTION
**Pull request by pazner**
# New LOR Functionality

* New "integrated" basis type for vector finite elements
* Spectrally equivalent LOR spaces for H(curl) and H(div)
* User-friendly LOR discretization and solver classes
* New LOR solver miniapp for all spaces: H1, H(curl), H(div), L2
* "Partially conforming" LOR discretizations on AMR meshes:
    * Spectrally equivalent LOR solvers for high-order problems on AMR meshes

Minor changes:
* Makes everything in `QuadratureFunctions1D` static
* Adds some convenience functions to `Geometry` for tensor-product geometries
* Add const accessor to `OperatorHandle`
* Fix some bugs in `BasisType`
* Implements `GetTransferMatrix` based on L2 projection when using the new basis with AMR meshes

## Some information about the LOR classes

(This is also available as documentation for users in `miniapps/solvers/lor_docs.md`).

This PR introduces some new LOR classes intended to make working with low-order refined discretizations and solvers more convenient.
The interface is designed to be convenient, flexible, and powerful: solvers with "default" options can be created in only one line of code, but much more customization is available if desired.

### Creating solvers with default options

The most basic LOR solver can be created as a one liner. For example, the following creates a direct solver based on a serial LOR discretization:

```c++
// Given BilinearForm `a` and list of essential DOFs, create direct LOR solver
LORSolver<UMFPackSolver> prec(a, ess_tdof_list);
```
This will automatically create a LOR discretization corresponding to `a`, assemble the system, and create a solver whose type is given by the template parameter.
For example, in parallel, one could do the following to create an AMG preconditioner instead of direct solver:
```c++
LORSolver<HypreBoomerAMG> prec(a, ess_tdof_list);
```

#### What does `LORSolver` do?

In the above examples, `LORSolver<T>` does a couple of things:

* Creates a LOR mesh and finite element space
* Assembles a LOR version of the bilinear form `a`
* Creates a preconditioner of type `T` using the LOR-assembled version
* Creates a DOF permutation (if required) to transition between the LOR and high-order DOF numbering

### Creating discretizations

If one is interested in using the underlying LOR discretization, it can be created as follows:
```c++
LOR lor(a, ess_tdof_list);
SparseMatrix &A_lor = lor.GetAssembledMatrix();
```
The low-order refined finite element space can be accessed by `lor.GetFESpace()`.
In parallel, there is an analogous `ParLOR` class:
```c++
ParLOR lor(a, ess_tdof_list);
HypreParMatrix &A_lor = lor.GetAssembledMatrix();
```

### Passing options to solvers

Some solvers require special options passed to the constructor.
For example, this can be useful when creating AMS preconditioners, which require access to the finite element space:
```c++
ParLOR lor(a, ess_tdof_list);
LORSolver<HypreAMS> ams(lor, &lor.GetParFESpace());
```
In this case, parameters passed to the `LORSolver` constructor will be forwarded to the `HypreAMS` constructor.
Also, note that the underlying solver object can be accessed with the `*` and `->` operators.

### Assembling custom bilinear forms

In all of the above examples, the LOR discretization is based off of the high-order bilinear form: the same form is assembled on a LOR finite element space.
In some circumstances, it is desirable to assemble a custom bilinear form on the LOR space and *not* to reuse the high-order form.
For example, one may want to assemble several LOR bilinear forms, and avoid the overhead of creating a mesh and finite element space for each of them.
These capabilities can be accomplished with the `LOR` and `LORSolver` classes:
```c++
LOR lor(fes); // Create LOR version of fes, don't assemble any forms
BilinearForm a_lor(&lor.GetFESpace());
a_lor.AddDomainIntegrator(new MassIntegrator);
a_lor.Assemble();
LORSolver<DSmoother> D(a_lor.SpMat(), lor);
```
In this case, the object `lor` does not assemble any forms, and does not represent any particular operator.
Instead, it simply represents a LOR mesh and finite element space, which can be used to create custom bilinear forms.
Once assembled, these operators can be passed to `LORSolver`, which will take care of any low-order-to-high-order DOF permutations, if needed.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2089](https://github.com/mfem/mfem/pull/2089) | @pazner | @tzanio | @barker29 + @dylan-copeland + @tzanio | 04/27/21 | 05/16/21 | 05/19/21 | |

### CODE REVIEW

**Comment by barker29** [comment]
Just so you know, I will be holding up this review until #1781 is merged into master, just so I can say, for a few days, that I wrote the best high-order H(curl) solver in MFEM.

More seriously, thanks for doing this and I'm looking forward to seeing how it performs.

**Comment by tzanio** [comment]
`make style` :)

**Comment by crdohrm** [comment]
Wow! Looks like a lot got done since our last discussion and my getting back from vacation today. Based on a quick look at the commits, it seems like you have both the degree of freedom and sign (+1 or -1) maps between the HO and LOR discretizations. Do I have that correct? Looking forward to hearing how this works out.

**Comment by pazner** [comment]
> Wow! Looks like a lot got done since our last discussion and my getting back from vacation today. Based on a quick look at the commits, it seems like you have both the degree of freedom and sign (+1 or -1) maps between the HO and LOR discretizations. Do I have that correct? Looking forward to hearing how this works out.

Yes, I think we've got the DOF numbering and orientations figured out now. Thanks for your help with that! We'll keep you in the loop as we progress with this

**Review by barker29 — COMMENTED** [review]
This is already very impressive. I'm seeing quite nice convergence and fast implementation at high order.

I am probably jumping the gun providing a "review" of a WIP branch, but I was excited to try this out and wanted to report what I noticed.

Can you suggest a command line for `plor` where `--hybridization` works? I can't get it to converge.

File: fem/fe.hpp
Comment (pazner) [reply fem/fe.hpp:41]: Yeah, I wasn't so happy with the name myself. Think of it as a placeholder for now. I'm not sure `Gerritsma` is any better, since it's not standard terminology, and you'd have to read the paper anyway to know what it means.

We could try to think of another term that suggests that instead of a *nodal* basis, it is a basis that satisfies a sort of "integral indicator function" condition. The other thing to note is that it is also dependent on the partitioning of the reference interval, which in this case is Gauss-Lobatto.

If we wanted to go for maximum verbosity, we could have `GaussLobattoIntegralIndicator`, which conveys the meaning but it awkward to read and write.

**Comment by pazner** [comment]
> This is already very impressive. I'm seeing quite nice convergence and fast implementation at high order.
> 
> I am probably jumping the gun providing a "review" of a WIP branch, but I was excited to try this out and wanted to report what I noticed.
> 
> Can you suggest a command line for `plor` where `--hybridization` works? I can't get it to converge.

Thanks taking a look at this! Unfortunately hybridization is not working, and it's not clear (from a linear algebra perspective, not from an implementation perspective) how to get it to work well in this case. If you have any great ideas I'd be happy to hear them :)

**Comment by crdohrm** [comment]
I'm happy to see that work is progressing right long. I was wondering about peoples' thoughts on how this is working out so far. If everything is working properly, I would expect a sparse direct solver for the LOR system to work well as a preconditioner for the HO system based on the spectral equivalence of the two systems. Have you observed this for H(curl) and H(div) problems? Also, have you looked into using AMS or ADS preconditioners for the LOR system as a preconditioner for the HO system? I'm hoping that this all works out and will open the door for the investigation of other preconditioning strategies as well (I like both multigrid and domain decomposition). Thanks for all the nice work.

**Comment by pazner** [comment]
> I'm happy to see that work is progressing right long. I was wondering about peoples' thoughts on how this is working out so far. If everything is working properly, I would expect a sparse direct solver for the LOR system to work well as a preconditioner for the HO system based on the spectral equivalence of the two systems. Have you observed this for H(curl) and H(div) problems? Also, have you looked into using AMS or ADS preconditioners for the LOR system as a preconditioner for the HO system? I'm hoping that this all works out and will open the door for the investigation of other preconditioning strategies as well (I like both multigrid and domain decomposition). Thanks for all the nice work.

Good to hear from you! Yes, so far everything looks like it's working well, both with sparse direct solvers (in serial) as well as AMS and ADS preconditioners for the LOR system. We are also currently looking into other preconditioning strategies, hopefully we'll have some updates soon!

**Comment by crdohrm** [comment]
That's great news! Looking forward to hearing more as updates become available.

File: fem/bilinearform.hpp
Comment (pazner) [inline fem/bilinearform.hpp:179]: This is not (currently) used in this PR, but it could be useful in the future

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: miniapps/solvers/lor_docs.md
Comment (barker29) [inline miniapps/solvers/lor_docs.md:25]: Maybe most users won't care, but at least I am interested in how the Gerritsma basis functions are handled here. If the high-order form uses the standard MFEM basis, can you do anything with LOR? If so, how does it work?

File: fem/lor.hpp
Comment (barker29) [inline fem/lor.hpp:128]: I don't like the name `LOR` because I think it will be obscure to users and hard to google. I would prefer `LowOrderRefined`, or maybe `LORDiscretization` or `LORSystem`.

File: miniapps/solvers/README
Comment (barker29) [inline miniapps/solvers/README:67]: Should we mention Clark Dohrmann's paper here (or in `lor_docs.md`)?

File: miniapps/solvers/lor_docs.md
Comment (barker29) [inline miniapps/solvers/lor_docs.md:27]:
```suggestion
* Creates a DOF permutation (if required) to transition between the LOR and high-order DOF numbering
```

File: miniapps/solvers/plor_solvers.cpp
Comment (barker29) [inline miniapps/solvers/plor_solvers.cpp:34]: Forget my other comments, this is probably the place to cite Dohrmann's paper.

**Review by barker29 — COMMENTED** [review]
Unfortunately I do not have the time to give this PR the review it deserves, but based on what I've seen I'm excited about it and think it looks very good. I do think a bit more documentation should make clear to users that they can't just expect `LORSolver` to work magic on H(curl) with the standard basis functions - the miniapp is clear but the classes themselves less so.

File: miniapps/solvers/lor_docs.md
Comment (pazner) [reply miniapps/solvers/lor_docs.md:25]: This is a good point (which, incidentally, @v-dobrev also brought up today).

One option is to warn the user if the "wrong" basis type is being used. Another option is to apply a change of basis operation in the solver action the same way the DOF permutation is currently applied. I'll have to look more into both of these options to figure out the best way to go.

File: miniapps/solvers/plor_solvers.cpp
Comment (pazner) [reply miniapps/solvers/plor_solvers.cpp:34]: Good idea. Done.

File: fem/fe.cpp
Comment (dylan-copeland) [inline fem/fe.cpp:1480]: In the nodal case, this function just does interpolation, so why introduce L2 projection? Alternatively, could we do something similar to the `ProjectIntegrated` functions, which do integrated interpolation? This would be specialized in classes such as `ND_QuadrilateralElement`, which would call its own function instead of this base class function. I am not saying we definitely should do this, but some questions are (i) why do L2 projection vs. interpolation; (ii) is it clear to the user whether projection or interpolation is done in each case; (iii) is L2 projection too expensive or unstable for very high order? My initial thoughts are (i) L2 projection is more accurate, but it is not clear why this choice is made here other than it is general for various derived classes; (ii) it is not clear to the user, and mixing projection and interpolation in different cases may cause confusion or problems; (iii) for moderate order (with PA we are limited to about order 5 or 6 for 3D Nedelec anyway), the dense matrix inversion should be fine, but for high enough order the inversion may get expensive and possibly inaccurate/unstable.

File: fem/lor.hpp
Comment (dylan-copeland) [inline fem/lor.hpp:277]: If I understand this correctly, for a `LORSolver *s` pointer, `s->Mult` would call `LORSolver::Mult`; for a `LORSolver s` object, `s->Mult` would call `s.solver.Mult`. If that is correct, it seems perhaps convenient but also potentially confusing and likely to cause bugs (which may be hard to track down, i.e. it may run but not apply the permutation). Would it be clearer and safer to just have a `GetSolver()` function?

File: fem/fe.hpp
Comment (dylan-copeland) [reply fem/fe.hpp:41]: I agree that we should have a slightly more descriptive name, e.g. `IntegratedGLL`, `IntegratedTensor`, or `IntegratedGerritsma`.

Comment (dylan-copeland) [reply fem/fe.hpp:41]: I agree that we should have a slightly more descriptive name, e.g. `IntegratedGLL`, `IntegratedTensor`, or `IntegratedGerritsma`.

File: miniapps/solvers/lor_solvers.cpp
Comment (dylan-copeland) [inline miniapps/solvers/lor_solvers.cpp:31]: No space in SuiteSparse.

File: miniapps/solvers/lor_solvers.cpp
Comment (dylan-copeland) [inline miniapps/solvers/lor_solvers.cpp:29]: The Dorhmann paper refers to the refined mesh as GLL. Is there a reason for the different name here?

File: miniapps/solvers/lor_solvers.cpp
Comment (dylan-copeland) [inline miniapps/solvers/lor_solvers.cpp:186]: Perhaps add `if (visualization)`, so the user can decide whether to produce these files. Since ParaView is more rarely used than glvis (at least by me), it may also be convenient to have a ParaView output parameter, false by default, but I don't feel strongly about it.

File: miniapps/solvers/lor_solvers.cpp
Comment (pazner) [reply miniapps/solvers/lor_solvers.cpp:29]: I suppose Gauss-Lobatto, Gauss-Legendre-Lobatto (GLL), and Legendre-Gauss-Lobatto are all names for the same thing. Perhaps in abbreviations, people like to use GLL for Lobatto, so that GL can be used for Gauss-Legendre without ambiguity.

If we are spelling out the name in full, I think Gauss-Lobatto is fine. It's also consistent with MFEM's `BasisType`.

File: miniapps/solvers/README
Comment (dylan-copeland) [inline miniapps/solvers/README:76]: Typo

File: miniapps/solvers/lor_mms.hpp
Comment (dylan-copeland) [inline miniapps/solvers/lor_mms.hpp:71]: It would be more efficient to factor out `1 + 3*pi2`.

File: fem/lor.hpp
Comment (pazner) [reply fem/lor.hpp:277]: The purpose of overloading the `->` operator is to allow for setting solver-specific parameters, e.g.
```
LORSolver<HypreBoomerAMG> lor_solver(…);
lor_solver->SetRelaxType(…);
```

Of course, if you call `lor_solver->Mult` in this case, it will do the wrong thing, which may be a risk for confusion.

File: miniapps/solvers/lor_mms.hpp
Comment (dylan-copeland) [inline miniapps/solvers/lor_mms.hpp:50]: Since we had to do some work on the projection to support inhomogeneous BC, it seems better to use an exact solution like that in `ex3p` to ensure something doesn't break in the future.

File: miniapps/solvers/plor_solvers.cpp
Comment (dylan-copeland) [inline miniapps/solvers/plor_solvers.cpp:52]: Since the exact solution is for the Maxwell problem, the H(div) test does not converge as the order is increased. This could be confusing to users. We could add an exact solution for the H(div) case, or just not output the error in that case.

File: fem/fe.cpp
Comment (dylan-copeland) [inline fem/fe.cpp:1388]: Extra line

File: fem/fe.hpp
Comment (dylan-copeland) [inline fem/fe.hpp:1989]: The `Eval` functions could use some brief documentation, at least to describe the arguments.

**Comment by dylan-copeland** [comment]
If the miniapps will be part of the autotests, then I think the new LOR solver miniapps are sufficient tests, so that we don't need comprehensive unit tests for individual components of the new LOR code. I don't think it is really necessary, but it would still be nice to have tests to verify spectral equivalence.

File: miniapps/solvers/plor_solvers.cpp
Comment (pazner) [reply miniapps/solvers/plor_solvers.cpp:52]: There should be a different forcing term in the case of H(div). Are you really not seeing convergence? What commands are you running and what errors do you get?

File: miniapps/solvers/lor_mms.hpp
Comment (pazner) [reply miniapps/solvers/lor_mms.hpp:50]: Not sure what you mean by this.

File: miniapps/solvers/plor_solvers.cpp
Comment (dylan-copeland) [reply miniapps/solvers/plor_solvers.cpp:52]: I see now that there is a different forcing term for H(div), but I don't see convergence for this test. Also, the solution plot looks weird on some parts of the boundary.
mpirun -np 4 plor_solvers -m ../../data/fichera.mesh -fe r
-o 3
L2 error: 0.0964843 
-o 4
L2 error: 0.0929569

File: miniapps/solvers/lor_mms.hpp
Comment (pazner) [reply miniapps/solvers/lor_mms.hpp:50]: You can also run with a different geometry, which should allow you to test nonzero tangential components.

File: fem/fe.cpp
Comment (dylan-copeland) [reply fem/fe.cpp:1480]: Thanks, this makes sense. It seems we agree that possible future work (after this PR) may be to add special cases of integrated interpolation for some derived classes to replace the general L2 projection from the base class, and to improve the dense matrix inversion for high order.

File: miniapps/solvers/plor_solvers.cpp
Comment (pazner) [reply miniapps/solvers/plor_solvers.cpp:52]: Good catch, there was a bug in `RT_HexahedronElement::ProjectIntegrated`. Can you check if it's working properly for you now?

File: fem/lor.hpp
Comment (dylan-copeland) [reply fem/lor.hpp:128]: I would also prefer a more descriptive name like `LowOrderRefined`.

File: fem/lor.hpp
Comment (dylan-copeland) [reply fem/lor.hpp:277]: So if we replace this function with `GetSolver`, then `lor_solver->SetRelaxType(…);` becomes `lor_solver.GetSolver()->SetRelaxType(…);`. I think that would be safer.

File: fem/fe.hpp
Comment (pazner) [reply fem/fe.hpp:41]: My vote would be for `IntegratedGLL`.

File: miniapps/solvers/plor_solvers.cpp
Comment (dylan-copeland) [reply miniapps/solvers/plor_solvers.cpp:52]: Yes, it works now.

File: fem/lor.hpp
Comment (pazner) [reply fem/lor.hpp]: Ok, I removed the `->` operator, and added `GetSolver` (which returns a reference, rather than a pointer, since the underlying solver can never be NULL), so that now it will be `lor_solver.GetSolver().SetRelaxType(…);`

File: fem/fe.hpp
Comment (pazner) [reply fem/fe.hpp]: It now returns `"Integrated Gauss-Lobatto indicator"`

File: fem/fe.hpp
Comment (pazner) [reply fem/fe.hpp]: I changed it to `IntegratedGLL`. Let me know if that looks good to you.

File: fem/fe.hpp
Comment (pazner) [reply fem/fe.hpp:1993]: I added a Doxygen comment here explaining what it is doing.

File: fem/lor.hpp
Comment (pazner) [reply fem/lor.hpp]: I changed it to `LORDiscretization` (which goes well with the related class, `LORSolver`). `LowOrderRefinedDiscretization` is probably too verbose.

File: miniapps/solvers/lor_docs.md
Comment (pazner) [reply miniapps/solvers/lor_docs.md:25]: I think the best thing to do is to automatically apply a change of basis, but doing that properly and efficiently is somewhat involves, and will probably happen in a subsequent PR.

**Review by dylan-copeland — APPROVED** [review]
Thanks @pazner! This PR is not only very useful but also well written and user-friendly, which is challenging given the complexity of the new capabilities.

File: miniapps/solvers/lor_docs.md
Comment (barker29) [reply miniapps/solvers/lor_docs.md:25]: It's fine for that to be a follow-on PR, but I think the documentation should be very clear about when a user can expect good performance from an LOR solver (IMO it's not quite clear enough right now).

Comment (pazner) [reply miniapps/solvers/lor_docs.md:33]: I updated the documentation explaining which bases are spectrally equivalent, and I added code that checks this at runtime and emits a warning when non-spectrally equivalent bases are used.

**Review by crdohrm — COMMENTED** [review]
Did you want to say anything about the BasisType::GaussLobatto requirement for L2 spectral equivalence?

**Comment by pazner** [comment]
> Did you want to say anything about the BasisType::GaussLobatto requirement for L2 spectral equivalence?

The DG case is a bit more complicated. For the mass matrix, `BasisType::GaussLegendre` should be fine, and for the "stiffness matrix" one really wants to consider something like an IPDG discretization, which is a bit more involved.

**Review by barker29 — APPROVED** [review]
I'm very happy with how this turned out.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
This deserves some coverage in `CHANGELOG` 😁

File: miniapps/solvers/lor_mms.hpp
Comment (tzanio) [inline miniapps/solvers/lor_mms.hpp:1]: I know this is a simple file, but lets still add the COPYRIGHT header

File: miniapps/solvers/lor_solvers.cpp
Comment (tzanio) [inline miniapps/solvers/lor_solvers.cpp:40]:
```suggestion
//      High Order Methods for Partial Differential Equations. (2010)
```

File: miniapps/solvers/lor_solvers.cpp
Comment (tzanio) [inline miniapps/solvers/lor_solvers.cpp:48]:
```suggestion
//
// Compile with: make lor-solvers
//
// Sample runs:
```

File: miniapps/solvers/lor_solvers.cpp
Comment (tzanio) [inline miniapps/solvers/lor_solvers.cpp:167]: This whole miniapp is very clean. I really like it!

File: miniapps/solvers/lor_solvers.cpp
Comment (tzanio) [inline miniapps/solvers/lor_solvers.cpp:141]:
```suggestion
   // TODO: L2 diffusion not implemented with partial assembly
```

File: miniapps/solvers/plor_solvers.cpp
Comment (tzanio) [inline miniapps/solvers/plor_solvers.cpp:11]: See the comments in the serial version

**Review by tzanio — APPROVED** [review]
Really awesome and powerful stuff 🎉 

Thanks @pazner and @crdohrm !

File: miniapps/solvers/makefile
Comment (pazner) [reply miniapps/solvers/makefile:25]: I think it's fine to keep it. The miniapp executable and source file are still named `block-solvers`.

Comment (tzanio) [reply miniapps/solvers/makefile:25]: ok

**Comment by v-dobrev** [comment]
I'm seeing a build error from this PR when using out-of-source build with GNU make -- `make all` fails due to this: https://github.com/mfem/mfem/blob/4fa50723b938945dc772bb4c95641ce2191a6e62/miniapps/solvers/makefile#L48-L49
For out-of-source builds to work, the dependency should be `$(SRC)lor_mms.hpp`.

I wonder why the autotests did not catch this. Maybe the Gitlab CI will catch this?

**Comment by tzanio** [comment]
I don't know if the out-of-source GNU make builds are actually tested. We should check that and add it to https://github.com/mfem/mfem/issues/1255.

If the above is the only fix needed, lets just add it in the mfem-4.3-dev PR.

**Comment by pazner** [comment]
Sorry about that! I'm pretty sure this is not caught by the GitLab tests. Is that the only change that is needed?

**Comment by v-dobrev** [comment]
I think the change I mentioned above is the only thing needed -- we should double check of course. :-)

One option could be to switch the Mac builds in the CI to use out-of-source builds.

**Comment by tzanio** [comment]
Addressed in 6bad77ca147a60d2c86d4fa4a16bd4ebbd440569

**Pull request merged**
